### PR TITLE
abstract BeaconSender and refactor integration tests

### PR DESCRIPTION
**Pull request by mar1n3r0**

### CODE REVIEW

File: it/sender_util.go [hunk 1]
```diff
@@ -2,246 +2,106 @@ package it
...
-		}
-	}
-}
-
-func httpPostFormOldStyle(params url.Values, client *http.Client, cnt int) {
-	uaStr := params.Get("user.agent")
-
-	if len(uaStr) == 0 {
-		uaStr = "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36"
+func processFile(file string) ([]url.Values, error) {
```
Comment (ivan-stankov-golang) [inline it/sender_util.go:13]: processFile is used by oldStyleBeaconSender. Move it to file old_style_beacon_sender.go

File: it/sender_util.go [hunk 2]
```diff
@@ -2,246 +2,106 @@ package it
...
-	// fmt.Println(params.Get("request_headers"))
-
-	var headers map[string][]string
-
-	unmrErr := json.Unmarshal([]byte(params.Get("request_headers")), &headers)
-
-	if unmrErr != nil {
-		fmt.Println("Bad headers JSON")
-		fmt.Println(unmrErr)
+func scanFile(fileName string) ([]url.Values, error) {
```
Comment (ivan-stankov-golang) [inline it/sender_util.go:43]: scanFile is used by newStyleBeaconSender move it to file new_style_beacon_sender.go

File: it/sender_util.go (hunk 1 above)
Comment (ivan-stankov-golang) [reply it/sender_util.go:13]: make if part of structure oldStyleBeaconSender

File: it/sender_util.go (hunk 2 above)
Comment (ivan-stankov-golang) [reply it/sender_util.go:43]: make it part of newStyleBeaconSender

File: it/sender_util.go [hunk 3]
```diff
@@ -2,246 +2,106 @@ package it
...
-	} else {
-		req.Header.Add("User-Agent", "")
+	var result []url.Values
+	for scanner.Scan() {
+		err = json.Unmarshal([]byte(scanner.Text()), &row)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse json[%s]: %w", []byte(scanner.Text()), err)
+		}
+		beaconData := row
+		result = append(result, makeUrlValues(beaconData))
```
Comment (ivan-stankov-golang) [inline it/sender_util.go:60]: the function makeUrlValues for newStyleBeaconSender is different that the one for oldStyleBeaconSender

File: it/sender_util.go [hunk 4]
```diff
@@ -2,246 +2,106 @@ package it
...
-				reqD.Set(bK, bV.(string))
-			}
-
-			valuesSlc = append(valuesSlc, reqD)
-		}
-
-		if err := scanner.Err(); err != nil {
-			log.Fatal(err)
-		}
+func extractBeaconPrefix(value string) string {
```
Comment (ivan-stankov-golang) [inline it/sender_util.go:100]: used for oldStyleBeaconSender. move it to file old_style_beacon_sender.go and make it part of struct oldStyleBeaconSender

File: it/sender_util.go [hunk 5]
```diff
@@ -2,246 +2,106 @@ package it
 
 import (
```
Comment (ivan-stankov-golang) [inline it/sender_util.go:3]: there should not be file sender_util.go

File: it/sender_util.go [hunk 6]
```diff
@@ -2,246 +2,106 @@ package it
...
-		req.Header.Add("User-Agent", uaStr[0])
-	} else {
-		req.Header.Add("User-Agent", "")
+	var result []url.Values
+	for scanner.Scan() {
+		err = json.Unmarshal([]byte(scanner.Text()), &row)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse json[%s]: %w", []byte(scanner.Text()), err)
+		}
+		beaconData := row
```
Comment (ivan-stankov-golang) [inline it/sender_util.go:59]: do not copy variable row to beaconData. Rename variable row to beaconData. Move 
```
var beaconData map[string]string
```
inside the for.

File: it/new_style_beacon_sender.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+	cityName := headers["Cf-Ipcity"][0]
+	req, _ := http.NewRequest("POST", fmt.Sprintf("http://%v:%v/beacon/catcher", b.host, b.port), strings.NewReader(params.Encode()))
+
+	if userAgent, ok := headers["User-Agent"]; ok {
+		req.Header.Add("User-Agent", userAgent[0])
+	} else {
+		req.Header.Add("User-Agent", "")
+	}
+
+	req.Header.Add("Cf-Ipcountry", countryCode)
```
Comment (ivan-stankov-golang) [inline it/new_style_beacon_sender.go:80]: new struct method makeCountryCode(params)

File: it/new_style_beacon_sender.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+	req.Header.Add("Cf-Ipcity", cityName)
```
Comment (ivan-stankov-golang) [inline it/new_style_beacon_sender.go:81]: new struct method makeCityName(params)

File: it/new_style_beacon_sender.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+		}
+		result = append(result, items...)
+	}
+	return result, nil
+}
+
+func (b *newStyleBeaconSender) httpPost(params url.Values) error {
+	var headers map[string][]string
+
+	err := json.Unmarshal([]byte(params.Get("request.headers")), &headers)
```
Comment (ivan-stankov-golang) [inline it/new_style_beacon_sender.go:65]: new struct method headers, err := parseHeaders(params)

File: it/new_style_beacon_sender.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+	if err != nil {
+		return fmt.Errorf("unable to parse json[%s]: %w", params, err)
+	}
+
+	countryCode := headers["Cf-Ipcountry"][0]
+	cityName := headers["Cf-Ipcity"][0]
+	req, _ := http.NewRequest("POST", fmt.Sprintf("http://%v:%v/beacon/catcher", b.host, b.port), strings.NewReader(params.Encode()))
+
+	if userAgent, ok := headers["User-Agent"]; ok {
```
Comment (ivan-stankov-golang) [inline it/new_style_beacon_sender.go:74]: new struct method makeCountryCode(headers)

**Comment by ceckoslab** [comment]
@mar1n3r0 it looks good to me. 

One thing that we could improve is to move a way of having 2 fixture sources: `old_style` and `new_style`

The reason why we have old and new style is because initially I used one data structure to save bundles of beacons on the file system (old style) but later I moved to another structure that more frequently used to store such type of events in the file system.

Having the old style doesn't bring value and it will be better if we move away from it and perhaps completely delete the test code for old style and delete the `it/data/old_style`.

It's completely fine if we delete all the files in `it/data/old_style` without porting them to the format we got in ``it/data/new_style`.

File: it/beacon_sender.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	return result
+}
+
+func (b *BeaconSender) httpPost(params url.Values) error {
+	headers, err := b.parseHeaders(params)
+	if err != nil {
+		return fmt.Errorf("parse headers error: %w", err)
+	}
+
+	req, _ := http.NewRequest("POST", b.sender.BuildUrl("/beacon/catcher"), strings.NewReader(params.Encode()))
```
Comment (ivan-stankov-golang) [inline it/beacon_sender.go:97]: handle err

File: it/beacon_sender.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+	if userAgent, ok := headers["User-Agent"]; ok {
```
Comment (ivan-stankov-golang) [inline it/beacon_sender.go:99]: extract makeUserAgent

File: it/beacon_sender.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+	return headers, nil
+}
+
+func (b *BeaconSender) makeCountryCode(headers map[string][]string) string {
+	countryCode := headers["Cf-Ipcountry"][0]
+	return countryCode
+}
+
+func (b *BeaconSender) makeCityName(headers map[string][]string) string {
+	countryCode := headers["Cf-Ipcity"][0]
```
Comment (ivan-stankov-golang) [inline it/beacon_sender.go:130]: directly return

**Comment by ivan-stankov-golang** [comment]
run `make lint`, which runs `golangci-lint run`

**Comment by ivan-stankov-golang** [comment]
extract migration logic from dao into MigrationDAO structure. 
Add prefix to migration tables.
```
	migrator := chmigrate.NewMigrator(db, migrations,
		chmigrate.WithTableName(p.prefix+"ch_migrations"),
		chmigrate.WithLocksTableName(p.prefix+"ch_migration_locks"),
	)
```

make lowercase MigrateDBURL and call it into the NewMigrationDAO constructor.

**Comment by ivan-stankov-golang** [comment]
@mar1n3r0 Looks good to me. Good work!

**Review by ceckoslab — APPROVED** [review]
LGTM!

**Pull request merged**